### PR TITLE
Fix language check on inserting code snippets into Script Editors

### LINKED ISSUE #1515: Inserting a code snippet to a script editor does not trigger language check
**Describe the issue**
The Code Snippets extension has the ability to check language compatibility of a snippet vs the editor the user wants to insert the snippet into, and display a warning message when there is a language mismatch.
The language check currently only works when inserting code snippets into a notebook cell.
image

This feature used to work with the Elyra Python Editor, but has been lost since the implementation of the new Script Editor for Python and R files.

**To Reproduce**
1. Open a Python or R Editor
2. Open the code snippets tab
3. Make sure you have a code snippet in a language different from the open editor
4. On the code snippet item, click the insert button
5. Notice the snippet is automatically added to the editor without displaying a warning dialog

**Expected behavior**
The user should see a warning dialog when adding a snippet of some language that does not match the open script editor.

**Deployment information**
Describe what you've deployed and how:
 - Elyra version: 2.2.1
 - Operating system: MacOs
 - Installation source: PyPI (using conda environment)
 - Browser: Chrome

### PR DESCRIPTION
**Pull request by karlaspuldaro**
Fixes #1515 
Restores language check for `ScriptEditor` widget when inserting a code snippet to the editor with the `insert` button

NOTE: Language check on code snippet drag and drop coming on a separate PR

Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

### CODE REVIEW

File: packages/code-snippet/src/CodeSnippetWidget.tsx [hunk 1]
```diff
@@ -163,6 +161,25 @@ class CodeSnippetDisplay extends MetadataDisplay<
...
+      widget instanceof DocumentWidget &&
+      (widget as DocumentWidget).content instanceof FileEditor
+    );
+  };
+
+  // Return the code editor from a given widget
+  private getFileEditor = (widget: Widget): CodeEditor.IEditor => {
+    const documentWidget = widget as DocumentWidget;
+    return (documentWidget.content as FileEditor).editor;
+  };
```
Comment (bourdakos1) [inline packages/code-snippet/src/CodeSnippetWidget.tsx:176]: You should be able to turn this into a type guard:
```ts
// type guard
const isFileEditor = (widget: Widget): widget is DocumentWidget<FileEditor> => {
  return (widget as DocumentWidget).content instanceof FileEditor;
};


if (isFileEditor(widget)) {
  // the widget should be automatically typed properly inside the if statement
  const fileEditor = widget.content.editor;
}
```

Comment (karlaspuldaro) [reply packages/code-snippet/src/CodeSnippetWidget.tsx]: Thank you @bourdakos1 , I like how cleaner it looks now

File: packages/code-snippet/src/CodeSnippetWidget.tsx [hunk 2]
```diff
@@ -104,28 +104,27 @@ class CodeSnippetDisplay extends MetadataDisplay<
...
-      (widget as DocumentWidget).content instanceof FileEditor
-    ) {
-      const documentWidget = widget as DocumentWidget;
-      const fileEditor = (documentWidget.content as FileEditor).editor;
+    if (this.isFileEditor(widget)) {
+      const fileEditor = widget.content.editor;
       const markdownRegex = /^\.(md|mkdn?|mdown|markdown)$/;
       if (
-        PathExt.extname(widget.context.path).match(markdownRegex) !== null &&
+        PathExt.extname((widget as DocumentWidget).context.path).match(
```
Comment (bourdakos1) [inline packages/code-snippet/src/CodeSnippetWidget.tsx:116]: this shouldn't need casting anymore

**Review by bourdakos1 — APPROVED** [review]
LGTM

File: packages/code-snippet/src/CodeSnippetWidget.tsx [hunk 3]
```diff
@@ -104,28 +104,25 @@ class CodeSnippetDisplay extends MetadataDisplay<
     this._evtMouseUp = this._evtMouseUp.bind(this);
   }
 
-  // Handle code snippet insert into an editor
+  // Handle code snippet insertion into an editor
   private insertCodeSnippet = async (snippet: IMetadata): Promise<void> => {
-    const widget: Widget = this.props.getCurrentWidget();
-    const snippetStr: string = snippet.metadata.code.join('\n');
+    const widget = this.props.getCurrentWidget();
+    const snippetStr = snippet.metadata.code.join('\n');
```
Comment (ajbozarth) [inline packages/code-snippet/src/CodeSnippetWidget.tsx:110]: is there a reason to make these less specific?

**Review by ajbozarth — COMMENTED** [review]
one nit, otherwise this lgtm code-wise, haven't tested it locally though

File: packages/code-snippet/src/CodeSnippetWidget.tsx (hunk 3 above)
Comment (bourdakos1) [reply packages/code-snippet/src/CodeSnippetWidget.tsx:110]: As long as there are inferred as `Widget` and `string` is there any reason for the annotation?

Comment (karlaspuldaro) [reply packages/code-snippet/src/CodeSnippetWidget.tsx:110]: Less verbose. And to follow the convention across the rest of the code, unless we really need to specify a type on variable declaration, like when calling some function that can return any type

Comment (ajbozarth) [reply packages/code-snippet/src/CodeSnippetWidget.tsx:110]: it's just a readability thing, I'm surprised this got past the linter though, usually it complains when things aren't typed. This isn't blocking or anything, just seemed like a code change for "no reason"

Comment (karlaspuldaro) [reply packages/code-snippet/src/CodeSnippetWidget.tsx:110]: If we want to type our variables, should we consider doing it consistently across all our code and packages?

Comment (ajbozarth) [reply packages/code-snippet/src/CodeSnippetWidget.tsx:110]: I thought the linter was already enforcing typed vars across our code base but it seem like it is not the case (unsure if this was changed or I just mis-implemented it)

Comment (bourdakos1) [reply packages/code-snippet/src/CodeSnippetWidget.tsx:110]: I think it only enforces it in areas that are more bug prone, like the return type of a function.

The typescript eslint plugin recommends using the rule `@typescript-eslint/no-inferrable-types`

Comment (ajbozarth) [reply packages/code-snippet/src/CodeSnippetWidget.tsx:110]: either way, if the linter doesn't mind this then I'm ok either way for now

Comment (bourdakos1) [reply packages/code-snippet/src/CodeSnippetWidget.tsx:110]: @ajbozarth you might want to take a look at https://github.com/elyra-ai/eslint-config-elyra, I'm trying to put together a shared eslint config if you have an rule suggestions

Comment (ajbozarth) [reply packages/code-snippet/src/CodeSnippetWidget.tsx:110]: I'll try to take a look, but why are we changing our lint rules exactly? I spent quite a while creating our current ones to match what the jupyter ecosystem uses. Or is that repo just moving the current lint rules into an external repo for multiple repos to share?

Comment (bourdakos1) [reply packages/code-snippet/src/CodeSnippetWidget.tsx:110]: the intention is to move current lint rules to external repo for all repos to share, but I'm using what I currently have set up for `pipeline-editor` so I'm not sure if it matches up 100% with what is in `elyra`

Comment (bourdakos1) [reply packages/code-snippet/src/CodeSnippetWidget.tsx:110]: the intention is to move current lint rules to external repo for all repos to share, but I'm using what I currently have set up for `pipeline-editor` so I'm not sure if it matches up 100% with what is in `elyra`

Comment (lresende) [reply packages/code-snippet/src/CodeSnippetWidget.tsx:110]: We should be consistent, but the "implementation details" are still being discussed and we should definitely use the one from Elyra as the base one.

**Comment by lresende** [comment]
I got confused with the discussions and wasn't sure if this is ready to merge. Please let me know.

**Comment by bourdakos1** [comment]
@lresende I believe this is ready to merge. We were just having a side conversation about linting

**Pull request merged**